crates/matrix-sdk/src/event_cache/deduplicator.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Simple but efficient types to find duplicated events. See [`Deduplicator`]
+//! to learn more.
+
+use std::{collections::BTreeSet, fmt, sync::Mutex};
+
+use growable_bloom_filter::{GrowableBloom, GrowableBloomBuilder};
+
+use super::room::events::{Event, RoomEvents};
+
+/// `Deduplicator` is an efficient type to find duplicated events.
+///
+/// It uses a [bloom filter] to provide a memory efficient probabilistic answer
+/// to: “has event E been seen already?”. False positives are possible, while
+/// false negatives are impossible. In the case of a positive reply, we fallback
+/// to a linear (backward) search on all events to check whether it's a false
+/// positive or not
+///
+/// [bloom filter]: https://en.wikipedia.org/wiki/Bloom_filter
+pub struct Deduplicator {
+    bloom_filter: Mutex<GrowableBloom>,
+}
+
+impl fmt::Debug for Deduplicator {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter.debug_struct("Deduplicator").finish_non_exhaustive()
+    }
+}
+
+impl Deduplicator {
+    const APPROXIMATED_MAXIMUM_NUMBER_OF_EVENTS: usize = 800_000;
+    const DESIRED_FALSE_POSITIVE_RATE: f64 = 0.001;
+
+    /// Create a new `Deduplicator`.
+    pub fn new() -> Self {
+        Self {
+            bloom_filter: Mutex::new(
+                GrowableBloomBuilder::new()
+                    .estimated_insertions(Self::APPROXIMATED_MAXIMUM_NUMBER_OF_EVENTS)
+                    .desired_error_ratio(Self::DESIRED_FALSE_POSITIVE_RATE)
+                    .build(),
+            ),
+        }
+    }
+
+    /// Scan a collection of events and detect duplications.
+    ///
+    /// This method takes a collection of events `events_to_scan` and returns a
+    /// new collection of events, where each event is decorated by a
+    /// [`Decoration`], so that the caller can decide what to do with these
+    /// events.
+    ///
+    /// Each scanned event will update `Self`'s internal state.
+    ///
+    /// `existing_events` represents all events of a room that already exist.
+    pub fn scan_and_learn<'a, I>(
+        &'a self,
+        events_to_scan: I,
+        existing_events: &'a RoomEvents,
+    ) -> impl Iterator<Item = Decoration<I::Item>> + 'a
+    where
+        I: Iterator<Item = Event> + 'a,
+    {
+        let mut already_seen = BTreeSet::new();
```

Comment:
1. Can you comment that this is a temporary variable (and not stored in `Deduplicator`) because we don't want an unbounded growth for this set? (and that it justifies the shape of the API, taking an in-parameter that's an iterator)
2. Can you also comment what its purpose is, here (i.e. find duplicates in the incoming iterator), please?